Support custom image/emoji favicons with `st.beta_set_favicon`

akrolsmir opened the pull request:
The harder half of #1006 

Note: Instead of using an emoji icon pack for emoji favicons, we just embed the system emoji as an SVG text node. This has the upside of keeping Streamlit slim, and the downside of a slightly smaller favicon that would otherwise be possible.

akrolsmir committed:
Create beta API to set page favicon

Addresses #1006

akrolsmir committed:
Allow emoji and emoji codes as favicons

akrolsmir committed:
Allow emojis to be directly passed to frontend

kantuni commented on frontend/src/components/core/Block/Block.tsx:
```diff
@@ -185,7 +186,7 @@ class Block extends PureComponent<Props> {
 
     const elementType = element.get("type")
-    const isEmpty = elementType === "empty"
+    const isEmpty = elementType === "empty" || elementType === "favicon"
```
I think the variable name should be changed also. `isEmptyOrFavicon`?

akrolsmir replied:
Renamed this (and the html class name) to `isHidden`

kantuni commented on frontend/src/components/elements/Favicon/Favicon.tsx:
```diff
@@ -0,0 +1,99 @@
+ * Hidden element that overwrites the page's favicon with the provided image
+ */
+export class Favicon extends Component<Props> {
+  emoji = ""
```
I believe it's better when access modifiers are transparent. Thoughts?

akrolsmir replied:
I actually switched to a functional implementation to get the component to re-render correctly -- these are now instances of useState

kantuni commented on frontend/src/components/elements/Favicon/Favicon.tsx:
```diff
@@ -0,0 +1,99 @@
+    ) : null
+  }
+
+  setFavicon(imageUrl: string) {
```
same here

akrolsmir replied:
Since this wasn't really a class method to begin with (didn't use any references of `this`), I've extracted it as a pure function.

kantuni commented on frontend/src/components/elements/Favicon/Favicon.tsx:
```diff
@@ -0,0 +1,99 @@
+  }
+
+  public componentDidMount() {
+    const svgElement = ReactDOM.findDOMNode(this) as Element
```
I would prefer to avoid `findDOMNode` if possible. Try using refs instead.
From React docs:
> `findDOMNode` is an escape hatch used to access the underlying DOM node. In most cases, use of this escape hatch is discouraged because it pierces the component abstraction. It has been deprecated in `StrictMode`.

akrolsmir replied:
Ah, good catch. useRef makes it cleaner and better typed, thanks.

kantuni commented on frontend/src/components/elements/Favicon/Favicon.tsx:
```diff
@@ -0,0 +1,99 @@
+import React, { Component, ReactNode } from "react"
+import ReactDOM from "react-dom"
+import { Map as ImmutableMap } from "immutable"
+import { getImageURI } from "../ImageList/ImageList"
```
We use a pattern of having `index.tsx` in each component folder, so that we can import by saying `import Foo from "/components/Foo"` (instead of `"/components/Foo/Foo"`)

kantuni commented on frontend/src/components/elements/ImageList/ImageList.tsx:
```diff
@@ -36,7 +36,7 @@ enum WidthBehavior {
 }
 
-function getImageURI(imgProto: ImmutableMap<string, any>): string {
+export function getImageURI(imgProto: ImmutableMap<string, any>): string {
```
This looks like a `utils.ts` function. But feel free to leave it here if you believe this is a better place.

akrolsmir replied:
Extracted as a  UriUtil function (and consolidated logic in Audio.tsx, Video.tsx too)

kantuni commented on lib/streamlit/DeltaGenerator.py:
```diff
@@ -1529,6 +1529,55 @@ def image(
+        """
+        from .elements import image_proto
+
+        width = -1  # Always use full width for favicons
```
Do we have a constant or enum for this?

akrolsmir replied:
Mm, not really. It's specified in the comments of Image.proto as well as the frontend implementation, but on the Python side there's no constant or enum, it's just a magic value (e.g. set as the default when width isn't specified for an ImageList).

I think the ideal solution is to change the -1/-2 from a commented solution into an enum in the proto so that Python and JS can point to the same value, but imo it's out scope for the PR.

kantuni reviewed (changes requested):
Nice work! Let me know if something is unclear.

akrolsmir committed:
Rename empty descriptor to hidden

akrolsmir committed:
Explicitly set visibility of class members

akrolsmir committed:
Extract media URI logic as util

akrolsmir committed:
Treat any error as emoji

Emoji might cause Invalid Argument or FileNotFoundException

akrolsmir committed:
Restructure Favicon as a functional component

This allows us to correctly rerender the favicon when the user updates their Streamlit function

akrolsmir committed:
Add a unit test for building media URIs

akrolsmir committed:
Add JS unit tests for the favicon

akrolsmir committed:
Add unit tests for emoji favicon

akrolsmir committed:
Test for changed props

kantuni reviewed (approved):
Great implementation! 👏 

The only comment would be to add return types to functions. 
P.S. Feel free to do it in a separate PR if you want to.

akrolsmir committed:
Add return types to all functions
